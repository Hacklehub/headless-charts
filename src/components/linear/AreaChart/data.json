[
  {
    "year": 2012,
    "macbook": 23.22,
    "services": 12.89,
    "wearables": 10.76,
    "ipad": 30.94,
    "iphone": 78.69
  },
  {
    "year": 2013,
    "macbook": 21.48,
    "services": 16.05,
    "wearables": 10.12,
    "ipad": 31.98,
    "iphone": 91.28
  },
  {
    "year": 2014,
    "macbook": 24.08,
    "services": 18.06,
    "wearables": 8.379,
    "ipad": 30.28,
    "iphone": 101.99
  },
  {
    "year": 2015,
    "macbook": 25.47,
    "services": 19.91,
    "wearables": 10.07,
    "ipad": 23.23,
    "iphone": 155.04
  },
  {
    "year": 2016,
    "macbook": 22.83,
    "services": 24.35,
    "wearables": 11.13,
    "ipad": 20.63,
    "iphone": 136.7
  },
  {
    "year": 2017,
    "macbook": 25.57,
    "services": 32.7,
    "wearables": 12.83,
    "ipad": 18.8,
    "iphone": 139.34
  },

  {
    "year": 2018,
    "macbook": 25.2,
    "services": 39.75,
    "wearables": 17.38,
    "ipad": 18.38,
    "iphone": 164.89
  },
  {
    "year": 2019,
    "macbook": 25.74,
    "services": 46.29,
    "wearables": 24.48,
    "ipad": 21.28,
    "iphone": 142.38
  },
  {
    "year": 2020,
    "macbook": 28.62,
    "services": 53.77,
    "wearables": 30.62,
    "ipad": 23.72,
    "iphone": 137.78
  },

  {
    "year": 2021,
    "macbook": 35.19,
    "services": 68.42,
    "wearables": 38.37,
    "ipad": 31.86,
    "iphone": 191.97
  },
  {
    "year": 2022,
    "macbook": 40.18,
    "services": 78.13,
    "wearables": 41.24,
    "ipad": 29.29,
    "iphone": 205.49
  }
]
