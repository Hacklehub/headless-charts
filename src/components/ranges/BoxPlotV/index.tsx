import { ZoomTransform, zoom } from 'd3-zoom';
import { axisBottom, axisLeft, axisRight, axisTop } from 'd3-axis';
import { max, min } from 'd3-array';
import { scaleBand, scaleLinear } from 'd3-scale';
import { select, selectAll } from 'd3-selection';
import useTooltip, { TooltipObjectType } from '../../../hooks/useTooltip';

import { ChartProps } from '../../../types';
import React from 'react';
import { defaultChartClassNames } from '../../../utils';
import { transition } from 'd3-transition';
import { twMerge } from 'tailwind-merge';

export interface BoxPlotVProps extends ChartProps {
  classNameData?: string;
  y: {
    minKey: string;
    maxKey: string;
    boxStart: string;
    boxEnd: string;
    midKey: string;
    min?: number;
    max?: number;
    axis?: 'right' | 'left';
    axisTicks?: number;
    classNameBoxes?: string;
    classNameLines?: string;
    classNameMap?: any;
  };
  x: {
    key: string;
    axis?: 'top' | 'bottom';
  };
  tooltip?: TooltipObjectType;
}

const BoxPlotV = ({
  className,
  classNameData,
  data,
  id,
  margin = {
    top: 40,
    bottom: 40,
    left: 40,
    right: 20,
  },
  padding = {
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
    bar: 0.2,
  },
  x,
  tooltip,
  y,
  zooming = {
    enabled: false,
  },
  style = {},
}: BoxPlotVProps) => {
  const { onMouseOver, onMouseMove, onMouseLeave } = useTooltip({
    id,
    tooltip,
    defaultHtml: (d: any) =>
      `min: ${d[y.minKey].toFixed(0)} <br/> range: ${+d[y.boxStart].toFixed(
        0
      )} to ${+d[y.boxEnd].toFixed(0)} <br/> mid: ${d[y.midKey].toFixed(
        0
      )} <br/> max: ${d[y.maxKey].toFixed(0)} `,
  });
  const refreshChart = React.useCallback(() => {
    const svg = select(`#${id}`);
    svg.selectAll('*').remove();

    const width = +svg.style('width').split('px')[0],
      height = +svg.style('height').split('px')[0];

    const g = svg.append('g');

    const xFn: any = scaleBand()
      .domain(data.map((d: any) => d[x.key]))
      .range([
        margin.left + (padding.left || 0),
        width - margin.right - (padding.right || 0),
      ])
      .padding(padding.bar || 0.2);

    const yFn = scaleLinear()
      .domain([
        Number.isFinite(y.min) ? y.min : min(data.map((d: any) => d[y.minKey])),
        Number.isFinite(y.max) ? y.max : max(data.map((d: any) => d[y.maxKey])),
      ])
      .range([
        height - margin.bottom - (padding.bottom || 0),
        (padding.top || 0) + margin.top,
      ]);

    svg
      .append('clipPath')
      .attr('id', 'clip')
      .append('rect')
      .attr('x', margin.left)
      .attr('y', margin.top - (padding.top || 0))
      .attr('width', width)
      .attr('height', height - margin.bottom - margin.top);

    const dataG = g
      .append('g')
      .attr('class', 'data')
      .attr('clip-path', 'url(#clip)');

    const dotRowsG = dataG
      .selectAll('g')
      .data(data)
      .enter()
      .append('g')
      .on('mouseenter', onMouseOver)
      .on('mousemove', onMouseMove)
      .on('mouseleave', onMouseLeave);

    transition();

    dotRowsG
      .append('line')
      .attr('clip-path', 'url(#clip)')
      .attr('x1', (d: any) => xFn(d[x.key]) + xFn.bandwidth() / 2)
      .attr('x2', (d: any) => xFn(d[x.key]) + xFn.bandwidth() / 2)
      .attr('y1', (d: any) => yFn(d[y.minKey]))
      .attr('y2', (d: any) => yFn(d[y.minKey]))
      .attr('class', `box-plot-line stroke-current ${classNameData || ``}`)
      .transition()
      .duration(1000)
      .attr('y2', (d: any) => yFn(d[y.maxKey]));

    // Begin lines
    dotRowsG
      .append('line')
      .attr('clip-path', 'url(#clip)')
      .attr('x1', (d: any) => xFn(d[x.key]))
      .attr('x2', (d: any) => xFn(d[x.key]))
      .attr('y1', (d: any) => yFn(d[y.minKey]))
      .attr('y2', (d: any) => yFn(d[y.minKey]))
      .attr('class', (d: any) =>
        twMerge('box-plot-line stroke-current', d.className || ``)
      )
      .transition()
      .duration(1000)
      .attr('x1', (d: any) => xFn(d[x.key]))
      .attr('x2', (d: any) => xFn(d[x.key]) + xFn.bandwidth());

    // Mid lines
    dotRowsG
      .append('line')
      .attr('clip-path', 'url(#clip)')
      .attr('y1', (d: any) => yFn(d[y.midKey]))
      .attr('y2', (d: any) => yFn(d[y.midKey]))
      .attr('x1', (d: any) => xFn(d[x.key]) + xFn.bandwidth() / 2)
      .attr('x2', (d: any) => xFn(d[x.key]) + xFn.bandwidth() / 2)
      .attr('class', `box-plot-line stroke-current ${classNameData || ``}`)
      .transition()
      .duration(1000)
      .attr('x1', (d: any) => xFn(d[x.key]))
      .attr('x2', (d: any) => xFn(d[x.key]) + xFn.bandwidth());

    // End lines
    dotRowsG
      .append('line')
      .attr('clip-path', 'url(#clip)')
      .attr('y1', (d: any) => yFn(d[y.maxKey]))
      .attr('y2', (d: any) => yFn(d[y.maxKey]))
      .attr('x1', (d: any) => xFn(d[x.key]) + xFn.bandwidth() / 2)
      .attr('x2', (d: any) => xFn(d[x.key]) + xFn.bandwidth() / 2)
      .attr('class', `box-plot-line stroke-current ${classNameData || ``}`)
      .transition()
      .duration(1000)
      .attr('x1', (d: any) => xFn(d[x.key]))
      .attr('x2', (d: any) => xFn(d[x.key]) + xFn.bandwidth());

    // Mid rects
    dotRowsG
      .append('rect')
      .attr(
        'class',
        twMerge(
          'box-plot-box stroke-current fill-current opacity-50 ',
          y.classNameBoxes
        )
      )
      .attr('clip-path', 'url(#clip)')
      .attr('y', () => height - margin.bottom - (padding.bottom || 0))
      .attr('x', (d: any) => xFn(d[x.key]))
      .attr('width', xFn.bandwidth())
      .transition()
      .duration(1000)
      .attr('y', (d: any) => yFn(d[y.boxEnd]))
      .attr('height', (d: any) => yFn(d[y.boxStart]) - yFn(d[y.boxEnd]));

    const yAxis = y.axis === 'right' ? axisRight(yFn) : axisLeft(yFn);

    const yAxisG = g
      .append('g')
      .attr('class', 'yAxis axis')
      .attr(
        'transform',
        `translate(${y.axis === 'right' ? margin.left + width : margin.left},0)`
      );
    yAxisG.call(yAxis);

    const xAxis = x.axis === 'top' ? axisTop(xFn) : axisBottom(xFn);

    const xAxisG = g.append('g').attr('class', 'axis--x axis');

    xAxisG
      .attr(
        'transform',
        `translate(0, ${
          x.axis === 'top' ? margin.top : height - margin.bottom
        })`
      )
      .call(xAxis);

    if (zooming?.enabled) {
      const zoomed = ({ transform }: { transform: ZoomTransform }) => {
        // transform g across only y axis
        dotRowsG.attr(
          'transform',
          `translate(0, ${transform.y}) scale(1, ${transform.k})`
        );
        yAxisG.call(yAxis.scale(transform.rescaleY(yFn)));
      };

      const zoomFn: any = zoom<SVGSVGElement, unknown>()
        .scaleExtent([zooming.min || 1, zooming.max || 2])
        .translateExtent([
          [0, 0],
          [width, height],
        ])
        .on('zoom', zoomed);

      svg.call(zoomFn);
    }
  }, [
    classNameData,
    data,
    id,
    margin,
    padding,
    x,
    y,
    zooming,
    onMouseLeave,
    onMouseMove,
    onMouseOver,
  ]);

  React.useEffect(() => {
    refreshChart();
    return () => {
      selectAll(`#tooltip-${id}`).remove();
    };
  }, [data, refreshChart]);
  return (
    <svg
      id={id}
      style={style}
      className={twMerge(defaultChartClassNames, className || '')}
    />
  );
};

export default BoxPlotV;
