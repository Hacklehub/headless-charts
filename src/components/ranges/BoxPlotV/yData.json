{
  "minKey": "min",
  "maxKey": "max",
  "midKey": "mid",
  "boxStart": "firstQuartile",
  "boxEnd": "lastQuartile",
  "min": 0
}
