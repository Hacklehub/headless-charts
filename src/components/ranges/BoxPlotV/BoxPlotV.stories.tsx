import BoxPlotV from '.';
import { Meta } from '@storybook/react';
import data from '../sample.json';

const meta: Meta<typeof BoxPlotV> = {
  title: 'Ranges/BoxPlotV',
  component: BoxPlotV,
  tags: ['autodocs'],
};

export default meta;

/** Default BoxPlot (Grouped). */
export const Default = {
  args: {
    data,
    id: 'box-plot-h-default',
    y: {
      minKey: 'min',
      maxKey: 'max',
      midKey: 'mid',
      boxStart: 'firstQuartile',
      boxEnd: 'lastQuartile',
      min: 0,
    },
    x: { key: 'name' },
  },
};

/** BoxPlot (Grouped) with custom colors. */
export const CustomColors = {
  args: {
    ...Default.args,
    id: 'box-plot-h-custom-colors',
    y: {
      ...Default.args.y,
      classNameBoxes: 'text-blue-500 opacity-100',
    },
  },
};

export const CustomColorMap = {
  args: {
    ...Default.args,
    id: 'box-plot-h-custom-color-map',
    y: {
      ...Default.args.y,
    },
    data: data.map((d: any, idx: number) => ({
      ...d,
      className: [
        'text-red-900',
        'text-green-900',
        'text-yellow-500',
        'text-orange-500',
      ][idx],
    })),
  },
};

export const DefaultTooltip = {
  args: {
    ...Default.args,
    id: 'box-plot-h-default-tooltip',
    tooltip: {},
  },
};

export const CustomStyleTooltip = {
  args: {
    ...Default.args,
    id: 'box-plot-h-custom-tooltip',
    tooltip: {
      className: 'bg-gray-100 text-gray-900',
    },
  },
};

export const CustomHtmlTooltip = {
  args: {
    ...Default.args,
    id: 'box-plot-h-custom-html-tooltip',
    tooltip: {
      html: (d: any) => {
        return `<div class="bg-gray-100 text-gray-900 p-2 rounded">
          <div class="text-lg font-bold">${d.name}</div>
          <div class="text-sm">${d.name} ranges between ${d.min} and ${d.max}, with an median of ${d.mid}. </div>
          <div class="text-sm">The first and last quartile values are at ${d.firstQuartile} and ${d.lastQuartile}</div>
        </div>`;
      },
    },
  },
};

export const Zooming = {
  args: {
    ...Default.args,
    id: 'box-plot-h-zooming',
    zooming: {
      enabled: true,
      min: 0.75,
      max: 2,
    },
  },
};
