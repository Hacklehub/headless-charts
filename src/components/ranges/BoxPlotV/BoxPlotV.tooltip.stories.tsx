/* eslint-disable @typescript-eslint/no-explicit-any */
import { Meta, StoryObj } from '@storybook/react';

import BoxPlotV from '.';
import data from '../sample.json';
import yData from './yData.json';

const meta: Meta<typeof BoxPlotV> = {
  title: 'Ranges/BoxPlotV/Tooltips',
  component: BoxPlotV,
  tags: ['autodocs'],
};

export default meta;

type Story = StoryObj<typeof BoxPlotV>;

export const DefaultTooltip: Story = {
  args: {
    data,
    id: 'box-plot-h-default-tooltip',
    tooltip: {},
    y: {
        ...yData
      },
      x: { key: 'name' },
  },
};

export const CustomStyleTooltip: Story = {
  args: {
    ...DefaultTooltip.args,
    id: 'box-plot-h-custom-tooltip',
    tooltip: {
      className: 'bg-gray-100 text-gray-900',
    },
  },
};

export const CustomHtmlTooltip: Story = {
  args: {
    ...DefaultTooltip.args,
    id: 'box-plot-h-custom-html-tooltip',
    tooltip: {
      html: (d: any) => {
        return `<div class="bg-gray-100 text-gray-900 p-2 rounded">
            <div class="text-lg font-bold">${d.name}</div>
            <div class="text-sm">${d.name} ranges between ${d.min} and ${d.max}, with an median of ${d.mid}. </div>
            <div class="text-sm">The first and last quartile values are at ${d.firstQuartile} and ${d.lastQuartile}</div>
          </div>`;
      },
    },
  },
};
