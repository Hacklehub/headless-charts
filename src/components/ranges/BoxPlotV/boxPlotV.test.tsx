import { describe, expect, it } from 'vitest';
import { render, screen } from '@testing-library/react';

import BoxPlotV from '.';
import data from '../sample.json';

describe('BoxPlotVChart', () => {
  it('renders boxplotV component for snapshot test', () => {
    render(
      <BoxPlotV
        data={data}
        id='boxPlotV'
        x={{ key: 'name' }}
        y={{
          minKey: 'min',
          maxKey: 'max',
          midKey: 'mid',
          boxStart: 'firstQuartile',
          boxEnd: 'lastQuartile',
          min: 0,
        }}
      />
    );
    const boxPlotV = screen.getByTestId('boxPlotV');
    expect(boxPlotV).toMatchSnapshot();
  });
  it('renders boxplotV component', () => {
    render(
      <BoxPlotV
        data={data}
        id='boxPlotV'
        x={{ key: 'name' }}
        y={{
          minKey: 'min',
          maxKey: 'max',
          midKey: 'mid',
          boxStart: 'firstQuartile',
          boxEnd: 'lastQuartile',
          min: 0,
        }}
      />
    );
    const boxPlotV = screen.getByTestId('boxPlotV');
    expect(boxPlotV).toBeInTheDocument();
  });
});
