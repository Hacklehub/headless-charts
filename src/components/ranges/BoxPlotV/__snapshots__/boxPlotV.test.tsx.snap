// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BoxPlotVChart > renders boxplotV component for snapshot test 1`] = `
<svg
  class="chart dark:bg-gray-800 dark:text-gray-50 h-64 lg:w-4/12 md:w-6/12 text-gray-900 w-full"
  data-testid="boxPlotV"
  id="boxPlotV"
>
  <g>
    <g
      class="data"
      clip-path="url(#clip)"
    >
      <g>
        <line
          class="box-plot-line stroke-current "
          clip-path="url(#clip)"
          x1="31.42857142857143"
          x2="31.42857142857143"
          y1="0"
          y2="0"
        />
        <line
          class="box-plot-line stroke-current"
          clip-path="url(#clip)"
          x1="25.714285714285715"
          x2="25.714285714285715"
          y1="0"
          y2="0"
        />
        <line
          class="box-plot-line stroke-current "
          clip-path="url(#clip)"
          x1="31.42857142857143"
          x2="31.42857142857143"
          y1="24"
          y2="24"
        />
        <line
          class="box-plot-line stroke-current "
          clip-path="url(#clip)"
          x1="31.42857142857143"
          x2="31.42857142857143"
          y1="40"
          y2="40"
        />
        <rect
          class=" box-plot-box fill-current opacity-50 stroke-current"
          clip-path="url(#clip)"
          width="11.428571428571429"
          x="25.714285714285715"
          y="-40"
        />
      </g>
      <g>
        <line
          class="box-plot-line stroke-current "
          clip-path="url(#clip)"
          x1="17.142857142857146"
          x2="17.142857142857146"
          y1="-40"
          y2="-40"
        />
        <line
          class="box-plot-line stroke-current"
          clip-path="url(#clip)"
          x1="11.42857142857143"
          x2="11.42857142857143"
          y1="-40"
          y2="-40"
        />
        <line
          class="box-plot-line stroke-current "
          clip-path="url(#clip)"
          x1="17.142857142857146"
          x2="17.142857142857146"
          y1="-24"
          y2="-24"
        />
        <line
          class="box-plot-line stroke-current "
          clip-path="url(#clip)"
          x1="17.142857142857146"
          x2="17.142857142857146"
          y1="-8"
          y2="-8"
        />
        <rect
          class=" box-plot-box fill-current opacity-50 stroke-current"
          clip-path="url(#clip)"
          width="11.428571428571429"
          x="11.42857142857143"
          y="-40"
        />
      </g>
      <g>
        <line
          class="box-plot-line stroke-current "
          clip-path="url(#clip)"
          x1="2.8571428571428603"
          x2="2.8571428571428603"
          y1="-24"
          y2="-24"
        />
        <line
          class="box-plot-line stroke-current"
          clip-path="url(#clip)"
          x1="-2.857142857142854"
          x2="-2.857142857142854"
          y1="-24"
          y2="-24"
        />
        <line
          class="box-plot-line stroke-current "
          clip-path="url(#clip)"
          x1="2.8571428571428603"
          x2="2.8571428571428603"
          y1="0"
          y2="0"
        />
        <line
          class="box-plot-line stroke-current "
          clip-path="url(#clip)"
          x1="2.8571428571428603"
          x2="2.8571428571428603"
          y1="24"
          y2="24"
        />
        <rect
          class=" box-plot-box fill-current opacity-50 stroke-current"
          clip-path="url(#clip)"
          width="11.428571428571429"
          x="-2.857142857142854"
          y="-40"
        />
      </g>
      <g>
        <line
          class="box-plot-line stroke-current "
          clip-path="url(#clip)"
          x1="-11.428571428571423"
          x2="-11.428571428571423"
          y1="-32"
          y2="-32"
        />
        <line
          class="box-plot-line stroke-current"
          clip-path="url(#clip)"
          x1="-17.14285714285714"
          x2="-17.14285714285714"
          y1="-32"
          y2="-32"
        />
        <line
          class="box-plot-line stroke-current "
          clip-path="url(#clip)"
          x1="-11.428571428571423"
          x2="-11.428571428571423"
          y1="-16"
          y2="-16"
        />
        <line
          class="box-plot-line stroke-current "
          clip-path="url(#clip)"
          x1="-11.428571428571423"
          x2="-11.428571428571423"
          y1="0"
          y2="0"
        />
        <rect
          class=" box-plot-box fill-current opacity-50 stroke-current"
          clip-path="url(#clip)"
          width="11.428571428571429"
          x="-17.14285714285714"
          y="-40"
        />
      </g>
    </g>
    <g
      class="yAxis axis"
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="end"
      transform="translate(40,0)"
    >
      <path
        class="domain"
        d="M-6,-39.5H0.5V40.5H-6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(0,-39.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,-31.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          10
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,-23.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          20
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,-15.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          30
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,-7.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          40
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,0.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          50
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,8.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          60
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,16.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          70
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,24.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          80
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,32.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          90
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,40.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          100
        </text>
      </g>
    </g>
    <g
      class="axis--x axis"
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="middle"
      transform="translate(0, -40)"
    >
      <path
        class="domain"
        d="M40.5,6V0.5H-19.5V6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(31.42857142857143,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          Type A
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(17.142857142857146,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          Type B
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(2.8571428571428603,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          Type C
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(-11.428571428571423,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          Type D
        </text>
      </g>
    </g>
  </g>
  <clippath
    id="clip"
  >
    <rect
      height="-80"
      width="0"
      x="40"
      y="40"
    />
  </clippath>
</svg>
`;
